addons/l10n_es_edi_sii/models/account_edi_format.py

```diff
@@ -0,0 +1,587 @@
+                    + tax_details_info_other_vals['tax_details']['base_amount']
+                    + tax_details_info_other_vals['tax_details']['tax_amount']
+                    - tax_details_info_other_vals['tax_amount_retention']
+                ), 2)
+
+                invoice_node['CuotaDeducible'] = round(sign * (
+                    tax_details_info_isp_vals['tax_amount_deductible']
+                    + tax_details_info_other_vals['tax_amount_deductible']
+                ), 2)
+
+            info_list.append(info)
+        return info_list
+
+    def _l10n_es_edi_web_service_aeat_vals(self, invoices):
+        if invoices[0].is_sale_document():
+            return {'url': 'https://www2.agenciatributaria.gob.es/static_files/common/internet/dep/aplicaciones/es/aeat/ssii_1_1/fact/ws/SuministroFactEmitidas.wsdl'}
+        else:
+            return {'url': 'https://www2.agenciatributaria.gob.es/static_files/common/internet/dep/aplicaciones/es/aeat/ssii_1_1/fact/ws/SuministroFactRecibidas.wsdl'}
+
+    def _l10n_es_edi_web_service_bizkaia_vals(self, invoices):
+        if invoices[0].is_sale_document():
+            return {
+                'url': 'https://www.bizkaia.eus/ogasuna/sii/documentos/SuministroFactEmitidas.wsdl',
+                'test_url': 'https://pruapps.bizkaia.eus/SSII-FACT/ws/fe/SiiFactFEV1SOAP',
+            }
+        else:
+            return {
+                'url': 'https://www.bizkaia.eus/ogasuna/sii/documentos/SuministroFactRecibidas.wsdl',
+                'test_url': 'https://pruapps.bizkaia.eus/SSII-FACT/ws/fr/SiiFactFRV1SOAP',
+            }
+
+    def _l10n_es_edi_web_service_gipuzkoa_vals(self, invoices):
+        if invoices[0].is_sale_document():
+            return {
+                'url': 'https://egoitza.gipuzkoa.eus/ogasuna/sii/ficheros/v1.1/SuministroFactEmitidas.wsdl',
+                'test_url': 'https://sii-prep.egoitza.gipuzkoa.eus/JBS/HACI/SSII-FACT/ws/fe/SiiFactFEV1SOAP',
+            }
+        else:
+            return {
+                'url': 'https://egoitza.gipuzkoa.eus/ogasuna/sii/ficheros/v1.1/SuministroFactRecibidas.wsdl',
+                'test_url': 'https://sii-prep.egoitza.gipuzkoa.eus/JBS/HACI/SSII-FACT/ws/fr/SiiFactFRV1SOAP',
+            }
+
+    def _l10n_es_edi_call_web_service_sign(self, invoices, info_list):
+        company = invoices.company_id
+        # All are sharing the same value, see '_get_batch_key'.
+        csv_number = invoices.mapped('l10n_es_edi_csv')[0]
+
+        # === Decode the certificate ===
+
+        certificate = invoices.company_id.l10n_es_edi_certificate_id
+        pem_certificate, pem_private_key, certificate = certificate._decode_certificate()
+        with NamedTemporaryFile(suffix='.cert') as cert_file, NamedTemporaryFile(suffix='.pem') as pem_file:
+            cert_file.write(pem_certificate)
+            cert_file.flush()
+            pem_file.write(pem_private_key)
+            pem_file.flush()
```

Comment:
This is not windows-compatible. I don't know if that is a concern.

Reply:
@xmo-odoo @william-andre Is there a way to make it?

Reply:
So you're dumping a plaintext private key in a temporary file that can be read by any process on the system? That's not good.

We've dealt with a similar case in #61853 where we wanted smtplib to use an SSL socket with a client cert, but loading the certificate and key from memory, not disk. We had to use `pyopenssl` because the stdlib did not support the necessary bindings.

Here the SSLContext is managed by `urllib3` via `requests`, and you already use a special `HTTPAdapter`, so you may not even need much, possibly just overriding `cert_verify()` in the adapter to set `cert_file` and `key_file` on the connection. Maybe this similar example can help:https://stackoverflow.com/a/63353645/1398110 (monkey patching may not be needed)

Reply:
If you can read files of the machine running the code, you can probably also access the database login credentials, and read the needed information from the database itself and then decode everything, can't you?

Reply:
But we will depend on pyopenssl again then.

Reply:
> If you can read files of the machine running the code, you can probably also access the database login credentials, and read the needed information from the database itself and then decode everything, can't you?

No, database credentials are not typically found in world-readable directories. And best practices include password-less connections to the database.

Reply:
> But we will depend on pyopenssl again then.

Yes, is that an issue? You can try-import it, and require it to be installed for this thing to work, no?

Reply:
It is just that @mart-e asked me to depend only on cryptography, but I guess it is not such a problem.

Reply:
> It is just that @mart-e asked me to depend only on cryptography, but I guess it is not such a problem.

Oh you can get it done with just `cryptography`, that's right, `pyopenssL` is just a wrapper above it, specialized in establishing SSL sockets. If you only need it to load certificates/keys, then yes, you should directly use `cryptography` (e.g. `cryptography.serialization`).

Reply:
@odony @smetl I did something here https://github.com/odoo-dev/odoo/commit/4c487a8104153925c646aeb7a114e3e67178f944 and it works, but somehow I did not get key_file to pass, so I passed the entire Odoo certificate object into the cert_file param and extended it all there.  Not sure also if it is healthy with all the overridden methods if somebody else needs it the normal way.  Also needed pyopenssl still.